incubator/hnc/internal/validators/namespace.go

```diff
@@ -64,48 +64,70 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
 // handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
 // tested (ie without constructing a full admission.Request).
 func (v *Namespace) handle(req *nsRequest) admission.Response {
-	parent := req.ns.Annotations[api.SubnamespaceOf]
-	if parent != "" && req.op == v1beta1.Delete {
-		msg := fmt.Sprintf("The namespace %s is a subnamespace. Please delete the anchor from the parent namespace %s to delete the subnamespace.", req.ns.Name, parent)
-		return deny(metav1.StatusReasonForbidden, msg)
-	}
-
-	return v.checkForest(req)
-}
-
-func (v *Namespace) checkForest(req *nsRequest) admission.Response {
 	v.Forest.Lock()
 	defer v.Forest.Unlock()
 
 	ns := v.Forest.Get(req.ns.Name)
 
-	// Allow all create namespace requests except if the namespace name already exists in
-	// the external hierarchy (the apiserver already denies existing namespace names).
-	if req.op == v1beta1.Create {
-		for _, nm := range v.Forest.GetNamespaceNames() {
-			if _, ok := v.Forest.Get(nm).ExternalTreeLabels[req.ns.Name]; ok {
-				msg := fmt.Sprintf("The namespace name %q is reserved by the external hierarchy manager %q.", req.ns.Name, v.Forest.Get(nm).Manager)
-				return deny(metav1.StatusReasonAlreadyExists, msg)
-			}
+	switch req.op {
+	case v1beta1.Create:
+		// Allow all create namespace requests except if the namespace name already exists in
+		// the external hierarchy (the apiserver already denies existing namespace names).
```

Comment:
This comment could get out of date ("allow all except in this one case") as soon as we add a second case. I'm not actually sure that we really need this comment at all, or the one on line 78 - it's pretty self-explanatory.